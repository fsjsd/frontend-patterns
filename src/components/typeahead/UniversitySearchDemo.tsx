import React, { useCallback } from 'react'
import styled from 'styled-components';

import { TypeAheadInput, TypeAheadResultsWrapper, TypeAheadResultItem, TypeAheadWrapper } from './TypeAheadStyles'
import TypeAhead from './TypeAhead';
import InputText from '../../ux/designsystem/InputText';
import Label from '../../ux/designsystem/Label';
import FieldSet from '../../ux/designsystem/FieldSet';

/**
 * UniversitySearchResult, based on types from the University Search API
 */
interface UniversitySearchResult {
  alpha_two_code: string;
  country: string;
  domains: string[];
  name: string;
  ["state-province"]: string | null;
  web_pages: string[];
}

const UniversitySearchInput = styled(InputText).attrs(() => ({
  placeholder: 'Search...'
}))``; // no styles here, just using attrs() to preset placeholder prop

/**
 * Results item for UniversitySearchResult display
 * @param param0 props
 * @returns JSX
 */
const UniversityResultItem = ({ datum, query, ...props }: { datum: UniversitySearchResult, query: string }) => {

  const queryLocation = datum.name.toLowerCase().indexOf(query);

  if (queryLocation !== -1) {
    const startText = datum.name.slice(0, queryLocation);
    const queryText = datum.name.slice(queryLocation, query.length + queryLocation);
    const endText = datum.name.slice(query.length + queryLocation);
    return <TypeAheadResultItem {...props}><b>{startText}</b>{queryText}<b>{endText}</b></TypeAheadResultItem>
  }

  return <TypeAheadResultItem {...props}>{datum.name}</TypeAheadResultItem>
}

/**
 * Simple UX component to show when no results are found
 * @returns JSX
 */
const NoResults = () => {
  return <TypeAheadResultItem>No results</TypeAheadResultItem>
}

const UniversitySearchDemo = () => {
  // straight forward fetch for University Search
  const handleGetResults = useCallback((query: string) => {
    return new Promise<UniversitySearchResult[]>((resolve) => {
      (async () => {
        const response = await fetch(`http://universities.hipolabs.com/search?name=${query}&country=australia`)
        const data = await response.json();
        resolve(data);
      })();
    })
  }, []);

  return (
    <FieldSet>
      <Label htmlFor="university-search">Search Australian Universities</Label>
      <TypeAhead
        id="university-search"
        wrapperComponent={TypeAheadWrapper}
        inputComponent={UniversitySearchInput}
        emptyResultsComponent={NoResults}
        resultsWrapperComponent={TypeAheadResultsWrapper}
        resultComponent={UniversityResultItem}
        getResults={handleGetResults}
        getResultText={(datum) => (datum as UniversitySearchResult).name}
      />
    </FieldSet>
  )
}

export default UniversitySearchDemo