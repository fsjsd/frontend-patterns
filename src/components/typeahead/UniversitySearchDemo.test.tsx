import React from 'react'
import { fireEvent, render, waitFor } from '@testing-library/react'
import UniversitySearchDemo from './UniversitySearchDemo';

describe('UniversitySearch', () => {
  globalThis.fetch = jest.fn(() =>
    Promise.resolve({
      json: () => Promise.resolve([
        {
          "state-province": null,
          "country": "Australia",
          "name": "University of Sydney",
          "web_pages": [
            "http://sydney.edu.au/"
          ],
          "domains": [
            "sydney.edu.au",
            "usyd.edu.au",
            "uni.sydney.edu.au"
          ],
          "alpha_two_code": "AU"
        },
        {
          "state-province": null,
          "country": "Australia",
          "name": "University of South Australia",
          "web_pages": [
            "http://www.unisa.edu.au/"
          ],
          "domains": [
            "unisa.edu.au"
          ],
          "alpha_two_code": "AU"
        },
        {
          "state-province": null,
          "country": "Australia",
          "name": "University of Southern Queensland",
          "web_pages": [
            "http://www.usq.edu.au/"
          ],
          "domains": [
            "usq.edu.au"
          ],
          "alpha_two_code": "AU"
        },
        {
          "state-province": null,
          "country": "Australia",
          "name": "University of Wollongong",
          "web_pages": [
            "http://www.uow.edu.au/"
          ],
          "domains": [
            "uow.edu.au"
          ],
          "alpha_two_code": "AU"
        },
        {
          "state-province": null,
          "country": "Australia",
          "name": "University of Western Australia",
          "web_pages": [
            "http://www.uwa.edu.au/"
          ],
          "domains": [
            "uwa.edu.au"
          ],
          "alpha_two_code": "AU"
        },
        {
          "state-province": null,
          "country": "Australia",
          "name": "University of Western Sydney",
          "web_pages": [
            "http://www.uws.edu.au/"
          ],
          "domains": [
            "uws.edu.au"
          ],
          "alpha_two_code": "AU"
        }
      ]),
    }),
  ) as jest.Mock;

  beforeEach(() => {
    (globalThis.fetch as jest.Mock).mockClear();
  });

  it('renders and matches snapshot', async () => {
    const { container, queryByLabelText } = render(<UniversitySearchDemo />)
    await waitFor(() => expect(queryByLabelText("Search Australian Universities")).toBeInTheDocument());
    expect(container).toMatchSnapshot();
  });

  it('typing less than 3 chars does nothing', async () => {
    const { container, getByRole } = render(<UniversitySearchDemo />)
    const input = await getByRole("textbox");
    fireEvent.change(input, { target: { value: "ab" } });
    expect(container).toMatchSnapshot();
  });

  it('valid query triggers result query', async () => {
    const { container, getByRole, queryByRole } = render(<UniversitySearchDemo />)
    const input = await getByRole("textbox");
    fireEvent.change(input, { target: { value: "University of" } });
    await waitFor(() => expect(queryByRole("listbox")).toBeInTheDocument());
    expect(container).toMatchSnapshot();
  });

})