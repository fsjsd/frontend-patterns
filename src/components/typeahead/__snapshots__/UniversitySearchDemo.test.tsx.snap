// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UniversitySearch renders and matches snapshot 1`] = `
<div>
  <label
    for="university-search"
  >
    Search Australian Universities
  </label>
  <div
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-owns="university-search-list"
    class="sc-bczRLJ hKJesX"
    role="combobox"
  >
    <input
      aria-autocomplete="list"
      aria-controls="typeahead-list"
      class="sc-gsnTZi sc-eCYdqJ hAsuur"
      id="university-search"
      placeholder="Search..."
      role="textbox"
      value=""
    />
  </div>
</div>
`;

exports[`UniversitySearch typing less than 3 chars does nothing 1`] = `
<div>
  <label
    for="university-search"
  >
    Search Australian Universities
  </label>
  <div
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-owns="university-search-list"
    class="sc-bczRLJ hKJesX"
    role="combobox"
  >
    <input
      aria-autocomplete="list"
      aria-controls="typeahead-list"
      class="sc-gsnTZi sc-eCYdqJ hAsuur"
      id="university-search"
      placeholder="Search..."
      role="textbox"
      value="ab"
    />
  </div>
</div>
`;

exports[`UniversitySearch valid query triggers result query 1`] = `
<div>
  <label
    for="university-search"
  >
    Search Australian Universities
  </label>
  <div
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-owns="university-search-list"
    class="sc-bczRLJ hKJesX"
    role="combobox"
  >
    <input
      aria-autocomplete="list"
      aria-controls="typeahead-list"
      class="sc-gsnTZi sc-eCYdqJ hAsuur"
      id="university-search"
      placeholder="Search..."
      role="textbox"
      value="University of"
    />
  </div>
</div>
`;
